File: load-balancer/java/src/main/java/com/danielwetan/softwareengineering/loadbalancer/LeastConnectionStrategy.java
package com.danielwetan.softwareengineering.loadbalancer;

import java.util.List;

public class LeastConnectionStrategy implements LoadBalancingStrategy {
  @Override
  public Server getServer(List<Server> servers, Request request) {
    int minConnections = Integer.MAX_VALUE;
    Server selectedServer = null;

    for (Server server : servers) {
      if (server.isHealthy()) {
        int connections = getConnections(server);
        if (connections < minConnections) {
          minConnections = connections;
          selectedServer = server;
        }
      }
    }
    return selectedServer;
  }

  private int getConnections(Server server) {
    // Perform logic to get current connections for the server
    return 0; // Placeholder for connections count
  }
}

File: load-balancer/go/least_connection_strategy.go
package main

import (
	"errors"
	"math"
)

var _ LoadBalancingStrategy = (*LeastConnectionStrategy)(nil)

type LeastConnectionStrategy struct{}

func (r *LeastConnectionStrategy) GetServer(servers []Server, request Request) (Server, error) {
	minConnections := math.MaxInt
	var selectedServer Server

	for _, server := range servers {
		if server.IsHealthy {
			connections := r.getConnections(server)

			if connections < minConnections {
				minConnections = connections
				selectedServer = server
			}
		}
	}

	if selectedServer.ServerID == "" {
		return Server{}, errors.New("no healthy servers available")
	}

	return selectedServer, nil
}

// Perform logic to get current connections for the server
func (r *LeastConnectionStrategy) getConnections(_ Server) int {
	return 0
}
